File: kakao.app/js/ked/view/colmns/dircol.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }}

var dircol

import kxk from "../../../kxk.js"
let post = kxk.post
let slash = kxk.slash

import belt from "../../edit/tool/belt.js"

import color from "../../theme/color.js"
import theme from "../../theme/theme.js"

import view from "../base/view.js"
import knob from "../base/knob.js"
import crumbs from "../base/crumbs.js"
import input from "../base/input.js"

import context from "../menu/context.js"

import dirtree from "./dirtree.js"


dircol = (function ()
{
    _k_.extend(dircol, view)
    function dircol (screen, editor, features)
    {
        this.editor = editor
    
        var bg, root

        this["onKey"] = this["onKey"].bind(this)
        this["onWheel"] = this["onWheel"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        this["onToggle"] = this["onToggle"].bind(this)
        this["onResize"] = this["onResize"].bind(this)
        this["rename"] = this["rename"].bind(this)
        this["onReveal"] = this["onReveal"].bind(this)
        this["onContextChoice"] = this["onContextChoice"].bind(this)
        this["onContext"] = this["onContext"].bind(this)
        this["draw"] = this["draw"].bind(this)
        this["layout"] = this["layout"].bind(this)
        this["setRoot"] = this["setRoot"].bind(this)
        this["onSessionMerge"] = this["onSessionMerge"].bind(this)
        this["onCrumbsAction"] = this["onCrumbsAction"].bind(this)
        dircol.__super__.constructor.call(this,screen,'dircol',features)
        this.isVisible = false
        this.active = true
        this.pointerType = 'pointer'
        this.knob = new knob(screen,`${this.name}_knob`)
        this.crumbs = new crumbs(screen,`${this.name}_crumbs`)
        this.dirtree = new dirtree(screen,`${this.name}_dirtree`,['scroll'])
        this.crumbs.on('action',this.onCrumbsAction)
        bg = theme.dirtree.bg
        this.dirtree.setColor('bg',bg)
        this.dirtree.setColor('empty',bg)
        this.dirtree.setColor('cursor_main',bg)
        this.dirtree.setColor('cursor_empty',bg)
        this.dirtree.scroll.setColor('bg',bg)
        this.knob.setColor('bg',bg)
        this.crumbs.setColor('empty',theme.gutter.bg)
        post.on('dircol.reveal',this.onReveal)
        post.on('dircol.resize',this.onResize)
        post.on('dircol.toggle',this.onToggle)
        post.on('dircol.root',this.setRoot)
        post.on('session.merge',this.onSessionMerge)
        root = ked_session.get('dircol▸root',process.cwd())
        this.setRoot(root)
    }

    dircol.prototype["onCrumbsAction"] = function (action, path)
    {
        if (action === 'click')
        {
            return this.setRoot(path)
        }
    }

    dircol.prototype["onSessionMerge"] = function (recent)
    {
        var args, root, _77_24_

        if (_k_.empty(recent.dircol))
        {
            return
        }
        args = ked_session.get('ked▸args',{})
        if (!_k_.empty(args.options))
        {
            console.log(`dircol.onSessionMerge - use first options dir ${slash.dir(args.options[0])}`)
            root = slash.dir(args.options[0])
        }
        else
        {
            console.log(`dircol.onSessionMerge - use last session dir ${recent.dircol.root}`)
            root = recent.dircol.root
        }
        if (root)
        {
            this.setRoot(root)
        }
        if ((recent.funcol != null ? recent.funcol.active : undefined))
        {
            this.active = true
            this.show()
        }
        return ked_session.set('dircol',recent.dircol)
    }

    dircol.prototype["setRoot"] = function (path)
    {
        if (_k_.empty(path))
        {
            return
        }
        path = slash.tilde(path)
        this.crumbs.set(path)
        this.dirtree.setRoot(path,{redraw:true})
        return ked_session.set('dircol▸root',path)
    }

    dircol.prototype["layout"] = function (x, y, w, h)
    {
        this.crumbs.layout(x,y,w,1)
        this.dirtree.layout(x,y + 1,w,h - 1)
        this.knob.layout(x + w - 1,y + 1,1,h - 1)
        return dircol.__super__.layout.call(this,x,y,w,h)
    }

    dircol.prototype["draw"] = function ()
    {
        if (this.hidden() || this.collapsed() || !this.active)
        {
            return
        }
        this.cells.fill_rect(0,1,-1,-1,' ',null,this.dirtree.color.bg)
        this.cells.fill_rect(0,0,-1,0,' ',null,this.crumbs.color.empty)
        this.crumbs.draw()
        this.dirtree.draw()
        this.knob.draw()
        return dircol.__super__.draw.call(this)
    }

    dircol.prototype["onContext"] = function (event)
    {
        return context.show(event.cell,this.onContextChoice,[" "," "," "])
    }

    dircol.prototype["onContextChoice"] = async function (choice)
    {
        var current, dir

        if (current = this.dirtree.current())
        {
            switch (choice)
            {
                case '':
                    return this.rename()

                case '':
                    dir = current.path
                    if (current.type === 'file')
                    {
                        dir = slash.dir(dir)
                    }
                    return post.emit('file.new_folder',dir)

                case '':
                    return post.emit('file.trash',current.path)

            }

        }
    }

    dircol.prototype["onReveal"] = async function (p)
    {
        var d, dd

        if (this.dirtree.itemForPath(p))
        {
            return
        }
        if (d = this.dirtree.itemForPath(slash.dir(p)))
        {
            if (!d.open)
            {
                return await this.dirtree.openDir(d,{redraw:true})
            }
            else
            {
                console.error(`dircol.onReveal already open? ${d}`)
            }
        }
        else
        {
            if (dd = this.dirtree.itemForPath(slash.dir(slash.dir(p))))
            {
                if (!dd.open)
                {
                    await this.dirtree.openDir(dd)
                    return await this.dirtree.openDir(d,{redraw:true})
                }
                else
                {
                    console.error(`dircol.onReveal already open dd? ${d}`)
                }
            }
            else
            {
                return this.setRoot(d)
            }
        }
    }

    dircol.prototype["rename"] = function ()
    {
        var current, fileName, ox, w, x, y

        current = this.dirtree.current()
        ox = belt.numIndent(current.tilde) + 2
        x = this.dirtree.cells.x + ox
        y = this.dirtree.cells.y + this.dirtree.currentIndex()
        w = this.dirtree.cells.cols - ox
        fileName = slash.file(current.path)
        return post.emit('input.popup',{text:fileName,x:x,y:y,w:w,cb:(function (res)
        {
            if (this.hover)
            {
                this.dirtree.grabFocus()
            }
            if (res !== fileName)
            {
                return post.emit('file.rename',current.path,slash.path(slash.dir(current.path),res))
            }
        }).bind(this)})
    }

    dircol.prototype["onResize"] = function ()
    {
        return this.knob.doDrag = true
    }

    dircol.prototype["onToggle"] = function ()
    {
        var cols

        if (!(this.visible() && this.collapsed()))
        {
            this.toggle()
        }
        this.active = this.visible()
        ked_session.set('dircol▸active',this.active)
        cols = _k_.max(16,parseInt(this.cells.screen.cols / 6))
        return post.emit('view.size',this.name,'right',((this.hidden() ? -this.cells.cols : cols - this.cells.cols)))
    }

    dircol.prototype["onMouse"] = function (event)
    {
        var ret

        if (this.hidden() || this.collapsed() || !this.active)
        {
            return
        }
        ret = dircol.__super__.onMouse.call(this,event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = this.knob.onMouse(event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = this.crumbs.onMouse(event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = this.dirtree.onMouse(event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
    }

    dircol.prototype["onWheel"] = function (event)
    {
        if (this.hidden() || this.collapsed() || !this.active)
        {
            return
        }
        return this.dirtree.onWheel(event)
    }

    dircol.prototype["onKey"] = function (key, event)
    {
        if (!this.dirtree.hasFocus())
        {
            return
        }
        switch (key)
        {
            case 'f2':
                return this.rename()

            case 'cmd+left':
            case 'ctrl+left':
                return this.setRoot(slash.dir(this.dirtree.currentRoot))

        }

        return this.dirtree.onKey(key,event)
    }

    return dircol
})()

export default dircol;

File: kakao.app/love/lua/view/colmns/dircol.lua
--[[
    ███████    ███  ████████    ███████   ███████   ███      
    ███   ███  ███  ███   ███  ███       ███   ███  ███      
    ███   ███  ███  ███████    ███       ███   ███  ███      
    ███   ███  ███  ███   ███  ███       ███   ███  ███      
    ███████    ███  ███   ███   ███████   ███████   ███████  
--]]

-- use ../../../kxk    ▪ post slash
-- use ../../edit/tool ◆ belt
-- use ../../theme     ◆ color theme
-- use ../base         ◆ view knob crumbs input
-- use ../menu         ◆ context
-- use                 ◆ dirtree

knob = require "view.base.knob"
view = require "view.base.view"
dirtree = require "view.colmns.dirtree"


local dircol = class("dircol", view)
    


function dircol:init(screen, editor, features) 
        self.editor = editor
        
        view.init(self, screen, 'dircol', features)
        
        self.isVisible = false
        self.active = true
        
        self.pointerType = 'pointer'
        
        self.knob = knob(screen, "" .. tostring(self.name) .. "_knob")
        self.crumbs = crumbs(screen, "" .. tostring(self.name) .. "_crumbs")
        self.dirtree = dirtree(screen, "" .. tostring(self.name) .. "_dirtree", array('scroll'))
        
        self.crumbs:on('action', self.onCrumbsAction)
        
        local bg = theme.dirtree.bg
        
        self.dirtree:setColor('bg', bg)
        self.dirtree:setColor('empty', bg)
        self.dirtree:setColor('cursor_main', bg)
        self.dirtree:setColor('cursor_empty', bg)
        
        self.dirtree.scroll:setColor('bg', bg)
        
        self.knob:setColor('bg', bg)
        
        self.crumbs:setColor('empty', theme.gutter.bg)
        
        post:on('dircol.reveal', self.onReveal)
        post:on('dircol.resize', self.onResize)
        post:on('dircol.toggle', self.onToggle)
        post:on('dircol.root', self.setRoot)
        
        post:on('session.merge', self.onSessionMerge)
        
        local root = ked_session:get('dircol▸root', slash.cwd())
        
        self:setRoot(root)
        return self
    end


function dircol:onCrumbsAction(action, path) 
        if (action == 'click') then 
            return self:setRoot(path)
        end
    end


function dircol:onSessionMerge(recent) 
        if empty(recent.dircol) then return end
        
        local args = ked_session.get('ked▸args', {})
        if valid(args.options) then 
            print("dircol.onSessionMerge - use first options dir " .. tostring(slash.dir(args.options[0])) .. "")
            local root = slash.dir(args.options[0])
        else 
            print("dircol.onSessionMerge - use last session dir " .. tostring(recent.dircol.root) .. "")
            local root = recent.dircol.root
        end
        
        if root then 
            self:setRoot(root)
        end
        
        if recent.funcol.active then 
            self.active = true
            self:show()
        end
        
        return ked_session.set('dircol', recent.dircol)
    end


function dircol:setRoot(path) 
        if empty(path) then return end
        
        path = slash.tilde(path)
        self.crumbs:set(path)
        self.dirtree:setRoot(path, {redraw = true})
        
        return ked_session:set('dircol▸root', path)
    end


function dircol:layout(x, y, w, h) 
        self.crumbs:layout(x, y, w, 1)
        self.dirtree:layout(x, (y + 1), w, (h - 1))
        self.knob:layout(((x + w) - 1), (y + 1), 1, (h - 1))
        
        view:layout(self, x, y, w, h)
        return self
    end


function dircol:draw() 
        if ((self:hidden() or self:collapsed()) or not self.active) then return end
        
        self.cells:fill_rect(0, 1, -1, -1, ' ', null, self.dirtree.color.bg)
        self.cells:fill_rect(0, 0, -1, 0, ' ', null, self.crumbs.color.empty)
        
        self.crumbs:draw()
        self.dirtree:draw()
        self.knob:draw()
        
        return super()
    end

--  ███████   ███████   ███   ███  █████████  ████████  ███   ███  █████████
-- ███       ███   ███  ████  ███     ███     ███        ███ ███      ███   
-- ███       ███   ███  ███ █ ███     ███     ███████     █████       ███   
-- ███       ███   ███  ███  ████     ███     ███        ███ ███      ███   
--  ███████   ███████   ███   ███     ███     ████████  ███   ███     ███   


function dircol:onContext(event) 
        return context:show(event.cell, self.onContextChoice, array(" ", " ", " "))
    end


function dircol:onContextChoice(choice) 
        local current = self.dirtree:current()
        if current then 
            if (choice == '') then 
                    return self:rename()
            elseif (choice == '') then 
                    local dir = current.path
                    if (current.type == 'file') then 
                        dir = slash.dir(dir)
                    end
                    
                    return post:emit('file.new_folder', dir)
            elseif (choice == '') then 
                    return post:emit('file.trash', current.path)
            end
        end
    end


function dircol:onReveal(p) 
        if self.dirtree:itemForPath(p) then return end
        local d = self.dirtree:itemForPath(slash.dir(p))
        if d then 
            if not d.open then 
                return self.dirtree:openDir(d, {redraw = true})
            else 
                return error("dircol.onReveal already open? " .. tostring(d) .. "")
            end
        else 
            local dd = self.dirtree:itemForPath(slash.dir(slash.dir(p)))
            if dd then 
                if not dd.open then 
                    self.dirtree:openDir(dd)
                    return self.dirtree:openDir(d, {redraw = true})
                else 
                    return error("dircol.onReveal already open dd? " .. tostring(d) .. "")
                end
            else 
                return self:setRoot(d)
            end
        end
    end

-- ████████   ████████  ███   ███   ███████   ██     ██  ████████
-- ███   ███  ███       ████  ███  ███   ███  ███   ███  ███     
-- ███████    ███████   ███ █ ███  █████████  █████████  ███████ 
-- ███   ███  ███       ███  ████  ███   ███  ███ █ ███  ███     
-- ███   ███  ████████  ███   ███  ███   ███  ███   ███  ████████


function dircol:rename() 
        local current = self.dirtree:current()
        local ox = (belt.numIndent(current.tilde) + 2)
        local x = (self.dirtree.cells.x + ox)
        local y = (self.dirtree.cells.y + self.dirtree:currentIndex())
        local w = (self.dirtree.cells.cols - ox)
        local fileName = slash.file(current.path)
        
        function cb(res) 
            if self.hover then self.dirtree:grabFocus() end
            if (res ~= fileName) then 
                return post.emit('file.rename', current.path, slash.path(slash.dir(current.path), res))
            end
        end
        
        return post:emit('input.popup', {text = fileName, x = x, y = y, w = w, cb = cb})
    end

-- █████████   ███████    ███████    ███████   ███      ████████
--    ███     ███   ███  ███        ███        ███      ███     
--    ███     ███   ███  ███  ████  ███  ████  ███      ███████ 
--    ███     ███   ███  ███   ███  ███   ███  ███      ███     
--    ███      ███████    ███████    ███████   ███████  ████████


function dircol:onResize() 
    self.knob.doDrag = true
    return self.knob.doDrag
    end

function dircol:onToggle() 
        if not (self:visible() and self:collapsed()) then self:toggle() end
        self.active = self:visible()
        ked_session:set('dircol▸active', self.active)
        local cols = max(16, int((self.cells.screen.cols / 6)))
        return -- post∙emit 'view.size' @name 'right' (@hidden() ? -@cells.cols : cols-@cells.cols )
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function dircol:onMouse(event) 
        if ((self:hidden() or self:collapsed()) or not self.active) then return end
        
        if view.onMouse(self, event) then return true end
        if self.knob:onMouse(event) then return true end
        if self.crumbs:onMouse(event) then return true end
        if self.dirtree:onMouse(event) then 
    return true
        end
    end

-- 000   000  000   000  00000000  00000000  000      
-- 000 0 000  000   000  000       000       000      
-- 000000000  000000000  0000000   0000000   000      
-- 000   000  000   000  000       000       000      
-- 00     00  000   000  00000000  00000000  0000000  


function dircol:onWheel(event) 
        if ((self:hidden() or self:collapsed()) or not self.active) then return end
        
        return self.dirtree:onWheel(event)
    end

-- 000   000  00000000  000   000  
-- 000  000   000        000 000   
-- 0000000    0000000     00000    
-- 000  000   000          000     
-- 000   000  00000000     000     


function dircol:onKey(key, event) 
        if not self.dirtree:hasFocus() then return end
        
        if (key == 'f2') then return self:rename()
        elseif (key == 'cmd+left') or (key == 'ctrl+left') then return self:setRoot(slash.dir(self.dirtree.currentRoot))
        end
        
        return self.dirtree:onKey(key, event)
    end

return dircol